model/riscv_insts_zfa.sail

```diff
@@ -0,0 +1,753 @@
+
+  write_fflags(fflags);
+  X(rd) = EXTZ(bool_to_bits(rd_val));
+  RETIRE_SUCCESS
+}
+
+/* FLTQ.D */
+
+union clause ast = RISCV_FLTQ_D : (regidx, regidx, regidx)
+
+mapping clause encdec = RISCV_FLTQ_D(rs2, rs1, rd)               if haveDExt() & haveZfa()
+  <-> 0b101_0001 @ rs2 @ rs1 @ 0b101 @ rd @ 0b101_0011           if haveDExt() & haveZfa()
+
+mapping clause assembly = RISCV_FLTQ_D(rs2, rs1, rd)
+  <-> "fltq.d" ^ spc() ^ freg_name(rd)
+               ^ spc() ^ freg_name(rs1)
+               ^ spc() ^ freg_name(rs2)
+
+function clause execute(RISCV_FLTQ_D(rs2, rs1, rd)) = {
+  let rs1_val_D = F_D(rs1);
+  let rs2_val_D = F_D(rs2);
+
+  let (fflags, rd_val) : (bits_fflags, bool) =
+      riscv_f64Lt_quiet   (rs1_val_D, rs2_val_D);
+
+  write_fflags(fflags);
+  X(rd) = EXTZ(bool_to_bits(rd_val));
+  RETIRE_SUCCESS
+}
+
+/* FCVTMOD.W.D */
+
+/*
+ * Implement float64 to int32 conversion without saturation, the
+ * result is supplied modulo 2^32.
+ * Rounding mode for FCVTMOD.W.D is always RTZ(Round toward zero).
+ */
+val      fcvtmod_helper : forall 'm, 'm == 64. (bits('m)) -> (bits(5), xlenbits)
+function fcvtmod_helper(x64) = {
+  let (sign, exp, mant) = fsplit_D(x64);
+  fflags		= fcsr.FFLAGS();
+  result : bits(64)	= undefined;
+
+  /* Handle special cases. */
+  if (exp == zeros())
+  then (nvFlag(), zeros())
+  /* NaN or Infinity. */
+  else if (exp == ones())
+  then (nvFlag(), zeros())
+  else {
+    let true_exp : int = unsigned(sub_vec_int(exp, 1023));
+    let shift    : int = true_exp - 52;
+    result	       = to_bits(64, unsigned(mant)) | ones() << 52;
+
+    /* Shift the fraction into place and set NX flag. */
+    if (shift >= 64 | shift <= -64) then {
+	/* The fraction is shifted out entirely. */
+	result = zeros();
+	fflags = fflags | nxFlag();
+    } else if (shift >= 0) then {
+	/* Raise NX if bit 52 got shifted out. */
+	result = result  << to_bits(64, shift);
+	if (unsigned(result) >= 12)
+	then fflags = fflags | nxFlag();
+    } else {
+	result = result >> to_bits(64, negate(shift));
+	/* Raise NX if bits got shifted out. */
+	if ((result << to_bits(64, negate(shift))) != result)
+	then fflags = fflags | nxFlag();
+    };
+
+    if (sign == 0b1)
+    then result = negate_D(result);
+
+    let result32b = to_bits(32, unsigned(result));
+    if (to_bits(64, unsigned(result32b)) != result)
+    then fflags = fflags | nvFlag();
+
+    (fflags, EXTS(result[31..0]))
+  };
+}
+
+union clause ast = RISCV_FCVTMOD_W_D : (regidx, regidx)
+
+/* We need rounding mode to be explicitly specified to RTZ(0b001) */
+mapping clause encdec = RISCV_FCVTMOD_W_D(rs1, rd)          if haveDExt() & haveZfa()
+  <-> 0b110_0001 @ 0b01000 @ rs1 @ 0b001 @ rd @ 0b101_0011  if haveDExt() & haveZfa()
+
+mapping clause assembly = RISCV_FCVTMOD_W_D(rs1, rd)
+  <-> "fcvtmod.w.d" ^ spc() ^ reg_name(rd)
+		    ^ spc() ^ freg_name(rs1)
+
+function clause execute(RISCV_FCVTMOD_W_D(rs1, rd)) = {
+  let (fflags, rd_val) = fcvtmod_helper(F_D(rs1));
```

Comment:
IMO it's better to keep the register read separate from the function on the result (i.e. revert the change you just made here since my review); that's what gets done consistently(?) elsewhere

Reply:
I thought that was the intent of your original comment... oh well, I seem to have overcompensated.
That change has been reverted (it was a separate commit anyway) and a new commit that drops empty lines has been added.